clippy_lints/src/inherent_impl.rs

```diff
@@ -56,33 +57,41 @@ impl<'tcx> LateLintPass<'tcx> for MultipleInherentImpl {
             ..
         }) = item.kind
         {
-            // Remember for each inherent implementation encountered its span and generics
+            // Remember for each inherent implementation encountered its id and span,
             // but filter out implementations that have generic params (type or lifetime)
             // or are derived from a macro
             if !in_macro(item.span) && generics.params.is_empty() {
-                self.impls.insert(item.def_id.to_def_id(), item.span);
+                self.impls.push((item.def_id, item.span));
             }
         }
     }
 
     fn check_crate_post(&mut self, cx: &LateContext<'tcx>, krate: &'tcx Crate<'_>) {
         if !krate.items.is_empty() {
-            // Retrieve all inherent implementations from the crate, grouped by type
-            for impls in cx.tcx.crate_inherent_impls(def_id::LOCAL_CRATE).inherent_impls.values() {
```

Comment:
Could we continue to use `crate_inherent_impls` and avoid *any* data on the lint pass?

Reply:
Both this change, and the change with storing the span have the downside of adding a table lookup for each impl block. This would avoid the vec allocation, which might make it better though. At the very least it's a little simpler.

Reply:
You would only need to lookup spans when you lint. Is there another query I'm missing? Generally I think it's better to not store what can be queried.

Reply:
There's a macro check and a check for generic parameters. Both could be held off on until the second matching type is seen, so not every impl block.

Reply:
> There's a macro check and a check for generic parameters.

Oh yeah I forgot that. Whatever you decide then.

Reply:
Tried switching to `crate_inherent_impls`, but it returns a hash map from the type definition to it's impl blocks. This causes the lint output to be in an undefined order which is both a small downgrade for people reading it, and can cause the tests to break for no reason.